llvm/lib/CheerpUtils/SIMDLowering.cpp

```diff
@@ -12,273 +12,805 @@
 #include "llvm/Cheerp/SIMDLowering.h"
 #include "llvm/Cheerp/GlobalDepsAnalyzer.h"
 #include "llvm/Cheerp/LinearMemoryHelper.h"
-#include "llvm/IR/Instruction.h"
-#include "llvm/IR/BasicBlock.h"
-#include "llvm/IR/InstIterator.h"
-#include "llvm/IR/Verifier.h"
+#include "llvm/IR/InstVisitor.h"
 #include "llvm/IR/IRBuilder.h"
 
 using namespace llvm;
 
-namespace cheerp
+struct VectorParts
 {
+	SmallVector<Value*, 16> values;
+
+	bool isNull()
+	{
+		return (values.size() == 0);
+	}
+};
 
-void SIMDLoweringPass::checkVectorCorrectness(Instruction& I)
+struct SIMDLoweringVisitor: public InstVisitor<SIMDLoweringVisitor, VectorParts>
 {
-	assert(isa<FixedVectorType>(I.getType()));
-	const FixedVectorType* vecTy = cast<FixedVectorType>(I.getType());
-	const unsigned vectorBitwidth = vecTy->getNumElements() * vecTy->getScalarSizeInBits();
-	if (vectorBitwidth != 128)
+	bool changed;
+	SmallVector<Instruction*, 16> toDelete;
+	SmallVector<PHINode*, 16> PHIs;
+	DenseMap<Value*, VectorParts> cache;
+	Type* Int32Ty;
+	Type* Int64Ty;
+	AllocaInst* bitcastAlloca;
+	const bool lowerAll;
+
+	SIMDLoweringVisitor(Module& M, bool lowerAll): changed(false), bitcastAlloca(nullptr), lowerAll(lowerAll)
+	{
+		Int32Ty = IntegerType::get(M.getContext(), 32);
+		Int64Ty = IntegerType::get(M.getContext(), 64);
+	}
+
+	~SIMDLoweringVisitor()
+	{
+		for (Instruction* I: toDelete)
+		{
+			I->replaceAllUsesWith(UndefValue::get(I->getType()));
+			I->eraseFromParent();
+		}
+		toDelete.clear();
+	}
+
+	bool shouldLower(const Type* type)
+	{
+		if (!type->isVectorTy())
+			return false;
+		const FixedVectorType* vecType = cast<FixedVectorType>(type);
+		const unsigned vectorSize = vecType->getNumElements() * vecType->getScalarSizeInBits();
+		if (vectorSize == 128 && !lowerAll)
+			return false;
+		if (vecType->getScalarSizeInBits() == 1)
```

Comment:
We should. I think this just never happened because then the auto-vectorization doesn't run.
The lowering for CmpInst (that will create vector of boolean) use custom logic, not this function, to decide whether they lower.

Reply:
Consider that we will always run auto-vectorization on the standard library, so some auto-vectorized code is espected to exist in any case